File: rust/y2015/d01/y2015d01.rs
pub fn part1(input: String) -> String {
    input
        .chars()
        .map(|p| match p {
            '(' => 1,
            ')' => -1,
            _ => 0,
        })
        .sum::<i64>()
        .to_string()
}

pub fn part2(input: String) -> String {
    let mut floor = 0;
    for (k, p) in input.chars().enumerate() {
        match p {
            '(' => floor += 1,
            ')' => floor -= 1,
            _ => panic!("!"),
        }
        if floor < 0 {
            return (k + 1).to_string();
        }
    }
    panic!("-1 not found")
}

File: pyaoc/solutions/y2015/y2015d01.py
from pyaoc import SolutionNotFound


def part1(input_str: str) -> str:
    value = {"(": 1, ")": -1}
    sol = sum(value[c] for c in input_str)
    return str(sol)


def part2(input_str: str) -> str:
    value = {"(": 1, ")": -1}
    floor = 0
    for k, c in enumerate(input_str):
        floor += value[c]
        if floor < 0:
            return str(k + 1)
    raise SolutionNotFound
